### PR TITLE
Various minor C++17 fixes and improvements

### PR DESCRIPTION
**Pull request by ZedThree**
Finally fixes a long standing warning coming from use of `std::tmpnam` when linking the unit tests. Replaced with `std::filesystem` and a class that deletes the file in its destructor.

Everything is pretty minor, and mostly just makes things a bit clearer and easier to read. There's a few places where we can use `if constexpr`, but I doubt we'll notice performance-wise.

### CODE REVIEW

**Comment by ZedThree** [comment]
Bah, we might want to look at running a more recent clang-format, this version is making some sub-optimal choices

**Review by bendudson — APPROVED** [review]
Thanks @ZedThree !

**Comment by ZedThree** [comment]
Looks like I actually broke something! Must've been building locally slightly differently

File: include/bout/petsc_interface.hxx [hunk 1]
```diff
@@ -435,9 +434,8 @@ public:
       const auto pws =
           pt->getWeightsForYApproximation(index2.x(), index1.y(), index2.z(), yoffset);
       const int ny =
-          std::is_same<T, FieldPerp>::value ? 1 : indexConverter->getMesh()->LocalNy;
-      const int nz =
-          std::is_same<T, Field2D>::value ? 1 : indexConverter->getMesh()->LocalNz;
+          std::is_same_v<T, FieldPerp> ? 1 : indexConverter->getMesh()->LocalNy;
+      const int nz = std::is_same_v<T, Field2D> ? 1 : indexConverter->getMesh()->LocalNz;
```
[the remark this answers is not in the document]
Comment (bendudson) [reply include/bout/petsc_interface.hxx:438]: This seems like an odd suggestion.

[the remark this answers is not in the document]
Comment (ZedThree) [reply include/bout/petsc_interface.hxx:438]: This must be a bug, the warning doesn't even make sense -- `nz` is definitely initialised!

**Review by bendudson — APPROVED** [review]

**Pull request merged**